src/io.rs

```diff
@@ -32,30 +40,147 @@ where
+    fn from(from: io::Error) -> RecordReadError {
+        RecordReadError::IoError { source: from }
+    }
+}
+
+/// A type for deserializing TFRecord formats
+#[derive(Debug)]
+pub struct RecordReader<R: Read> {
+    reader: R,
+}
+
+impl<R> RecordReader<R>
+where
+    R: Read,
+{
+    /// Construct a new RecordReader from an underlying Read.
+    pub fn new(reader: R) -> Self {
+        RecordReader { reader }
+    }
+
+    fn read_next_len(&mut self) -> Result<Option<u64>, RecordReadError> {
+        let len = match self.reader.read_u64::<LittleEndian>() {
+            Err(e) => {
+                if e.kind() == io::ErrorKind::UnexpectedEof {
+                    return Ok(None);
+                }
+                return Err(e.into());
+            }
+            Ok(val) => val,
+        };
+
+        let mut len_bytes = [0u8; 8];
+        LittleEndian::write_u64(&mut len_bytes, len);
+
+        let expected_len_crc32 = self.reader.read_u32::<LittleEndian>()?;
+        let actual_len_crc32 = mask(crc32::checksum_castagnoli(&len_bytes));
+        if expected_len_crc32 != actual_len_crc32 {
+            return Err(RecordReadError::InvalidLengthChecksum);
+        }
+        Ok(Some(len))
+    }
+
+    fn checksum_bytes(&mut self, bytes: &[u8]) -> Result<(), RecordReadError> {
+        let actual_bytes_crc32 = mask(crc32::checksum_castagnoli(&bytes));
+        let expected_bytes_crc32 = self.reader.read_u32::<LittleEndian>()?;
+        if actual_bytes_crc32 != expected_bytes_crc32 {
+            return Err(RecordReadError::InvalidContentChecksum);
+        }
+        Ok(())
+    }
+    fn read_bytes_exact(&mut self, len: u64, buf: &mut [u8]) -> Result<(), RecordReadError> {
+        if (buf.len() as u64) < len {
+            return Err(RecordReadError::BufferTooShort {
+                needed: len,
+                had: buf.len() as u64,
+            });
+        }
+        self.reader.read_exact(buf)?;
+        self.checksum_bytes(buf)?;
+        Ok(())
+    }
+    /// Read the next record into a byte slice.
+    /// Returns the number of bytes read, if successful.
+    /// Returns None, if it could read exactly 0 bytes (indicating EOF)
+    pub fn read_next(&mut self, buf: &mut [u8]) -> Result<Option<u64>, RecordReadError> {
+        let len = match self.read_next_len()? {
+            Some(len) => len,
+            None => return Ok(None),
+        };
+
+        let slice = &mut buf[0..len as usize];
+        self.read_bytes_exact(len, slice)?;
+
+        Ok(Some(len))
+    }
+    /// Allocate a Vec<u8> on the heap and read the next record into it.
+    /// Returns the filled Vec, if successful.
+    /// Returns None, if it could read exactly 0 bytes (indicating EOF)
+    pub fn read_next_owned(&mut self) -> Result<Option<Vec<u8>>, RecordReadError> {
+        let len = match self.read_next_len()? {
+            Some(len) => len,
+            None => return Ok(None),
+        };
+        let mut vec = vec![0u8; len as usize];
```

Comment:
I didn't bother with reserving capacity and unsafing to write into uninitialized memory, because I think that performance conscious users are going to use the other method anyway. But let me know what you think.